src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceWriter.cs

```diff
@@ -1354,6 +1396,29 @@ private void WriteRedSetters(Node node)
             }
         }
 
+        private TreeType GetBaseTypeWithField(TreeType node, string name)
+        {
+            TreeType bestType = null;
+            for (var current = node; current != null; current = TryGetBaseType(current))
+            {
+                var fields = GetNodeOrNodeListFields(current);
+                var field = fields.FirstOrDefault(f => f.Name == name);
+                if (field != null)
+                {
+                    bestType = current;
+                }
+            }
+
+            return bestType;
+        }
+
+        private TreeType TryGetBaseType(TreeType node)
+            => node is AbstractNode an
```

Comment:
Not related to this PR, but would have helped: should AbstractNode and Node have a common ancestor?